cursive-core/src/logger.rs

```diff
@@ -1,15 +1,71 @@
 //! Logging utilities.
 
 use lazy_static::lazy_static;
+use std::cmp::Ord;
 use std::collections::VecDeque;
-use std::sync::Mutex;
+use std::str::FromStr;
+use std::sync::{Mutex, RwLock};
 
 /// Saves all log records in a global deque.
 ///
 /// Uses a `DebugView` to access it.
+///
+/// # Examples
+///
+/// Set log levels from env vars
+///
+/// ```
+/// # use cursive_core::*;
+/// logger::set_filter_levels_with_env();
+/// logger::init();
+/// ```
+///
+/// Set log levels explicitly.
+///
+/// ```
+/// # use cursive_core::*;
+/// # use log::LevelFilter;
+/// logger::set_int_filter_level(LevelFilter::Warn);
+/// logger::set_ext_filter_level(LevelFilter::Debug);
+/// logger::init();
+/// ```
+
 pub struct CursiveLogger;
 
-static LOGGER: CursiveLogger = CursiveLogger;
+lazy_static! {
+    // Log filter level for log messages from within cursive
+    static ref INT_FILTER_LEVEL: RwLock<log::LevelFilter> = RwLock::new(log::LevelFilter::Trace);
+    // Log filter level for log messages from sources outside of cursive
+    static ref EXT_FILTER_LEVEL: RwLock<log::LevelFilter> = RwLock::new(log::LevelFilter::Trace);
+}
+
+/// Sets the internal log filter level.
+pub fn set_int_filter_level(level: log::LevelFilter) {
+    *INT_FILTER_LEVEL.write().unwrap() = level;
+}
+
+/// Sets the external log filter level.
+pub fn set_ext_filter_level(level: log::LevelFilter) {
+    *EXT_FILTER_LEVEL.write().unwrap() = level;
+}
+
+/// Sets log filter levels based on environment variables `RUST_LOG` and `CURSIVE_LOG`.
+/// If `RUST_LOG` is set, then both internal and external log levels are set to match.
+/// If `CURSIVE_LOG` is set, then the internal log level is set to match with precedence over
+/// `RUST_LOG`.
+pub fn set_filter_levels_with_env() {
+    if let Ok(rust_log) = std::env::var("RUST_LOG") {
+        if let Ok(filter_level) = log::LevelFilter::from_str(&rust_log) {
```

Comment:
I'd log (ha!) a warning in case we can't parse the level, rather than silently ignoring it.

Reply:
This is meant to be called before the logger is initialized, so any logged warning would be ignored. I guess you could call it after initialization, but then when initializing the `log::set_max_level()` would be based on the default filter levels (which is fine, cause it defaults to trace). I'll add it anyway.